emitter.py

```diff
@@ -72,49 +77,106 @@ def sanitize_payload(item, log, sanitize_func):
+        log.exception('http_emitter: Unable to convert message to json')
         # early return as we can't actually process the message
         return
     except RuntimeError as rte:
-        log.error('http_emitter: runtime error dumping message to json %s', rte)
+        log.exception('http_emitter: runtime error dumping message to json')
         # early return as we can't actually process the message
         return
     except Exception as e:
-        log.error('http_emitter: unknown exception processing message %s', e)
+        log.exception('http_emitter: unknown exception processing message')
         return
 
-    zipped = zlib.compress(payload)
+    for payload in payloads:
+        try:
+            headers = get_post_headers(agentConfig, payload)
+            r = requests.post(url, data=payload, timeout=5, headers=headers)
 
-    log.debug("payload_size=%d, compressed_size=%d, compression_ratio=%.3f"
-              % (len(payload), len(zipped), float(len(payload))/float(len(zipped))))
+            r.raise_for_status()
+
+            if r.status_code >= 200 and r.status_code < 205:
+                log.debug("Payload accepted")
 
+        except Exception:
+            log.exception("Unable to post payload.")
+
+
+def serialize_payload(message, log):
+    payload = ""
     try:
-        headers = get_post_headers(agentConfig, zipped)
-        r = requests.post(url, data=zipped, timeout=5, headers=headers)
+        payload = json.dumps(message)
+    except UnicodeDecodeError:
+        newmessage = sanitize_payload(message, log, remove_control_chars)
+        try:
+            payload = json.dumps(newmessage)
+        except UnicodeDecodeError:
+            log.info('Removing undecodable characters from payload')
+            newmessage = sanitize_payload(newmessage, log, remove_undecodable_chars)
+            payload = json.dumps(newmessage)
 
-        r.raise_for_status()
+    return payload
+
+
+def serialize_and_compress_legacy_payload(legacy_payload, max_compressed_size, depth, log):
+    """
+    Serialize and compress the legacy payload
+    """
+    serialized_payload = serialize_payload(legacy_payload, log)
+    zipped = zlib.compress(serialized_payload)
+    log.debug("payload_size=%d, compressed_size=%d, compression_ratio=%.3f"
+              % (len(serialized_payload), len(zipped), float(len(serialized_payload))/float(len(zipped))))
 
-        if r.status_code >= 200 and r.status_code < 205:
-            log.debug("Payload accepted")
 
-    except Exception:
-        log.exception("Unable to post payload.")
+    compressed_payloads = [zipped]
+
+    if len(zipped) > max_compressed_size:
+        # let's just log a warning for now, splitting the legacy payload is tricky
+        log.warning("collector payload is above the limit of %dKB compressed", max_compressed_size/(1<<10))
+
+    return compressed_payloads
+
+
+def serialize_and_compress_metrics_payload(metrics_payload, max_compressed_size, depth, log):
```

Comment:
This is a really cool, elegant solution.

However, it took me a bit to understand everything that was going on in it. I think it could stand to use a few comments. I think it would be very easy to make a mistake in editing this function in the future without some added clarity, and I specified some of the places where I think it could stand to be clearer in my other comments.